File: 백준/Gold/1759. 암호 만들기/암호 만들기.py
import sys
input = sys.stdin.readline

l, c = map(int, input().split())

array = sorted(input().split())
visited = [False]*c
# print(array)


def solve(password, idx):
    if len(password) == l:
        mocnt = 0
        for i in range(l):
            if password[i] in list('aeiou'):
                mocnt += 1

        if mocnt >= 1 and l-mocnt >= 2:
            print(''.join(password))
        return

    for i in range(idx, c):
        if not visited[i]:
            visited[i] = True
            password.append(array[i])
            solve(password, i+1)
            password.pop()
            visited[i] = False


for i in range(c-l+1):
    visited[i] = True
    solve([array[i]], i+1)
    visited[i] = False

File: 백준/Gold/1759. 암호 만들기/암호 만들기.js
const input = require("fs")
  .readFileSync("/dev/stdin")
  .toString()
  .trim()
  .split("\n")

const [l, c] = input[0].split(" ").map(Number)
const alphabet = input[1].split(" ").sort()
const visited = Array.from({ length: c }, () => false)
const 모음 = "aeiou"
const Composition = []

function dfs(array, startIdx) {
  if (array.length === l) {
    let moCnt = array.filter((alpha) => 모음.includes(alpha)).length // 모음 수
    let jaCnt = l - moCnt

    if (moCnt >= 1 && jaCnt >= 2) {
      Composition.push(array.join(''))
    }
    return
  }
  for (let i = startIdx; i < c; i++) {
    if (!visited[i]) {
      visited[i] = true
      dfs([...array, alphabet[i]], i)
      visited[i] = false
    }
  }
}

for (let i = 0; i < c; i++) {
  if (!visited[i]) {
    visited[i] = true
    dfs([alphabet[i]], i)
    visited[i] = false
  }
}
Composition.forEach((str)=>console.log(str))

